# A combinatorial domain screening platform reveals epigenetic effector interactions for transcriptional perturbation

Additional data and codes to accompany the preprint 


# Library 1


To address the issue of lower R2 sequencing quality, which is occasionally observed during paired-end readout of long fragments, we implemented an approach of spreading ten 15-mer short identifiers (IDs) across the entire 301 bp sequencing region. If any of the 10 IDs exactly matched the NGS read, the read was assigned to the corresponding domain. Unique 15 mer IDs for identifying N- and C-terminal domains were generated by the following steps. 


Step 1) From fasta files composed of expected 301 bp NGS reads for every domain (R1_Forward_P5_301cycles.fasta & R2_Reverse_P7_301cycles.fasta), all possible unique 15 mers were generated using UniqueKMER with the following commands. 


```bash
$ uniquekmer -f R1_Forward_P5_301cycles.fasta -k 15 -o "kmercollection_R1"
$ uniquekmer -f R2_Reverse_P7_301cycles.fasta -k 15 -o "kmercollection_R2"
```


Step 2) If a 15 mer ID for one domain had a hamming distance of less than 2 with an ID from another domain, one of the IDs was deleted from the domain with a larger number of IDs by executing Library1_step2.py. Output files were numerically sorted using Seqkit with the following commands.

'''bash
	$ seqkit sort -nN R1_15mer_HD2_UMIs.fasta -o R1_15mer_HD2_UMIs_sorted.fasta
	$ seqkit sort -nN R2_15mer_HD2_UMIs.fasta -o R2_15mer_HD2_UMIs_sorted.fasta
 '''
